src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -524,6 +538,24 @@ private static CallSiteResultCacheLocation GetCommonCacheLocation(CallSiteResult
             return parameterCallSites;
         }
 
+        /// <summary>
+        /// Verifies none of the generic type arguments are ValueTypes.
+        /// </summary>
+        /// <remarks>
+        /// NativeAOT apps are not guaranteed that the native code for the closed generic of ValueType
+        /// has been generated. To catch these problems early, this verification is enabled at development-time
+        /// to inform the developer early that this scenario will not work once AOT'd.
+        /// </remarks>
+        private static void VerifyOpenGenericAotCompatibility(Type serviceType, Type[] genericTypeArguments)
+        {
+            foreach (Type typeArg in genericTypeArguments)
+            {
+                if (typeArg.IsValueType)
```

Comment:
Is this really the only condition when the generic type cannot be created by NativeAOT?

Reply:
Yup - when the compiler sees a generic type being a target of reflection, it will create the canonical reference type instantiation for it. It is a heuristic that makes a lot of things work. We decided to make the heuristic contractual because it allows a lot of `MakeGeneric` to "just work", with the exact same perf characteristics as if it was statically predicted.

The only possible drawback of the heuristic is more code, but this only kicks in for targets of reflection and with `<TrimMode>full</TrimMode>` there's few of those.

Reply:
What about if `typeArg` isn't a value type, but is itself generic with a value type argument?

For example:

```cs
var typeArg = typeof(List<int>);
var newType = typeof(List<>).MakeGenericType(typeArg); // List<List<int>>
```

Does this succeed? I'm guessing yes, because the compiler can see `typeof(List<int>)` constant, and the list of list is still just a list over references, which can be shared.

Reply:
> I'm guessing this is ok because the compiler can see `typeof(List<int>)` constant, and the MakeGenericType call is for a list of lists, which is still just a list of references and can be shared.

Yes, that's correct.

Reply:
> The other part of the feature switch is that you get a consistent experience with and without AOT

I agree that such feature switch can add a value, but why do you think that every library or code developer should replicate such logic to achieve same/consistence experience with and without NativeAOT? I can imagine a scenario where developers are using JIT during development for some reasons and want to reduce a risk of breaking their app that ships like NativeAOT. Why do they need to rely on every caller of `MakeGenericType` to add extra check with feature switch instead of having one to enable? It'd be opt-out feature switch for sure but that's this one [too].(https://github.com/dotnet/runtime/pull/79425/files#diff-b9824a372eeae2c312b58b219349b8b0e4612d432c122f87b3ef45850ec2c7c2R38)

Reply:
I agree that it's not a solution we should ask everybody to use - it's just not practical that way.
Personally I think we'll have to introduce some kind of annotation, similar to the trimmer annotations, if this kind of pattern shows up a lot.

Right now I think the one big feature switch for all value types is too big of a hammer, but maybe it will turn out to be useful.

Reply:
We do the feature switch with JIT time emulation when we expect the situation is uncommon and we want to suppress static warnings.

I think `MakeGenericType` with valuetypes will always be a lot more common than DI with valuetypes.

Even if one goes through the MakeGenericType usage in the framework - most have the valuetype problem, unless we already gave them some extra AOT attention (like MakeGenericType in EventSource or DiagnosticSource that is now AOT safe).

If an API call is more likely to be an issue for AOT than a non-issue, a static warning is better than JIT time emulation.

Reply:
I agree that the API is "problematic" for NativeAOT but I think an alternative fix where we change `MakeGenericType` to give better context anyone when the type cannot be constucted, for example like

> You called MakeGenericType with value type argument of `Foo` under AOT setup, consider doing XYZ to resolve this problem.

would have bigger impact as it would help everyone that believes that their `MakeGenericType` call with value types will be used rarely and suppressing the warning is fine. As I understand the problem, it would also fix the problem in this PR where we are too aggressive with value type check because if the generic instance with value type is available/discovered elsewhere in the program `MakeGenericType` with value type argument should work or not?

Reply:
> As I understand the problem, it would also fix the problem in this PR where we are too aggressive with value type check because if the generic instance with value type is available/discovered elsewhere in the program MakeGenericType with value type argument should work or not?

I'm not sure I would call this a "fix" - unless such dependency is intentional it would make for a very fragile setup - somebody changes some part of the app, and a completely unrelated part suddenly breaks because the generic instantiation it was using is no longer statically referenced. Honestly even now relying on this behavior is problematic to day the least - I would much rather we have a declarative way of doing this, and one which is verifiable.

Reply:
> As I understand the problem, it would also fix the problem in this PR where we are too aggressive with value type check because if the generic instance with value type is available/discovered elsewhere in the program `MakeGenericType` with value type argument should work or not?

It is true that if the generic code exists in the app, the `MakeGenericType` would succeed with native AOT. But this behavior cannot be emulated with the JIT. We would have to allow this with any type on F5 debug. So we would be back to the behavior where things work with F5 debug, but fail when one is ready to ship. This behavior is not desirable for trimming, and it's not desirable for AOT either.

The mantra for AOT is the same as with trimming: the app works the same between F5 debug and PublishAot, **or** there's a warning.

We want to get rid of a warning here because 95+% of users won't be using DI with value types and a warning would be just noise.